data/trie/patriciaMerkleTrie.go

```diff
@@ -585,50 +591,36 @@ func (tr *patriciaMerkleTrie) GetSerializedNodes(rootHash []byte, maxBuffToSend
 	return nodes, remainingSpace, nil
 }
 
-// GetAllLeaves iterates the trie and returns a map that contains all leafNodes information
-func (tr *patriciaMerkleTrie) GetAllLeaves() (map[string][]byte, error) {
-	tr.mutOperation.RLock()
-	defer tr.mutOperation.RUnlock()
-
-	if tr.root == nil {
-		return map[string][]byte{}, nil
-	}
-
-	leaves := make(map[string][]byte)
-	err := tr.root.getAllLeaves(leaves, []byte{}, tr.Database(), tr.marshalizer)
-	if err != nil {
-		return nil, err
-	}
-
-	return leaves, nil
-}
-
 // GetAllLeavesOnChannel adds all the trie leaves to the given channel
-func (tr *patriciaMerkleTrie) GetAllLeavesOnChannel() chan core.KeyValueHolder {
+func (tr *patriciaMerkleTrie) GetAllLeavesOnChannel(rootHash []byte) (chan core.KeyValueHolder, error) {
 	//TODO pass a context for cancellation purposes when needed
-	leavesChannel := make(chan core.KeyValueHolder)
+	leavesChannel := make(chan core.KeyValueHolder, 100)
 
 	tr.mutOperation.RLock()
-	if tr.root == nil {
+
+	newTrie, err := tr.recreate(rootHash)
+	if err != nil || check.IfNil(newTrie) || tr.root == nil {
 		tr.mutOperation.RUnlock()
 		close(leavesChannel)
-
-		return leavesChannel
+		return leavesChannel, err
 	}
+
+	tr.EnterPruningBufferingMode()
 	tr.mutOperation.RUnlock()
 
 	go func() {
 		tr.mutOperation.RLock()
-		err := tr.root.getAllLeavesOnChannel(leavesChannel, []byte{}, tr.Database(), tr.marshalizer)
+		err = newTrie.root.getAllLeavesOnChannel(leavesChannel, []byte{}, tr.Database(), tr.marshalizer)
```

Comment:
shouldn't you return this error as well ? like errFound and mutex ?

Reply:
Will live like this at the moment, like discussed.